Content adjustment for upcoming onboarding steps

felixschlegel opened the pull request:
# Content adjustment for upcoming onboarding steps

## :recycle: Current situation & Problem

Closes #44.
`OnboardingStack` only allowed modifications to its views (i.e. order and content of the `OnboardingStack`) when the onboarding process was at its first step (i.e. no navigation operations have occurred). We want to loosen this restriction by allowing modifications to all views that are ahead of the current onboarding step.


## :gear: Release Notes 

* `OnboardingNavigationPath.updateViews`: update all views that are ahead of current onboarding step instead of only updating views on the first onboarding step

## :books: Documentation

* _No changes to public interface_

## :white_check_mark: Testing

- [x] Create UI Test

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

philippzagar commented:
@felixschlegel Thanks for the PR; it looks quite straightforward already! 🚀 

One random thought: We probably should get rid of "duplicate" `onboardingStepsOrder` array and `onboardingSteps` dictionary. The reason why I introduced it back then was for performance reasons (no need to iterate through the `onboardingStepsOrder` array when accessing a specific step), however, in retrospect, I'm not sure if that actually was the best way to solve this. Realistically speaking, the `OnboardingStack` contains just a handful of `View`s, therefore making the overhead of iterating through the onboarding array negligible. Maybe something like an [`OrderedDictionary `](https://github.com/apple/swift-collections/blob/main/Documentation/OrderedDictionary.md) would work out? 
Let me know what you think.

felixschlegel commented:
> @felixschlegel Thanks for the PR; it looks quite straightforward already! 🚀
> 
> One random thought: We probably should get rid of "duplicate" `onboardingStepsOrder` array and `onboardingSteps` dictionary. The reason why I introduced it back then was for performance reasons (no need to iterate through the `onboardingStepsOrder` array when accessing a specific step), however, in retrospect, I'm not sure if that actually was the best way to solve this. Realistically speaking, the `OnboardingStack` contains just a handful of `View`s, therefore making the overhead of iterating through the onboarding array negligible. Maybe something like an [`OrderedDictionary `](https://github.com/apple/swift-collections/blob/main/Documentation/OrderedDictionary.md) would work out? Let me know what you think.

Great point! Have adjusted the code accordingly.

Furthermore, I added a new UI Test assuring that steps ahead of the current onboarding steps can be modified.

PSchmiedmayer reviewed (approved):
@felixschlegel Thank you for the improvements; the changes look great to me. Thanks for adding the UI tests. Let me know once you think that we can merge the PR.

We have some unfortunate issues with codecov and are evaluating how we will continue with the tool beyond the issues here. I can merge the PR with admin permissions once you think it is complete 👍

felixschlegel commented:
> @felixschlegel Thank you for the improvements; the changes look great to me. Thanks for adding the UI tests. Let me know once you think that we can merge the PR.
> 
> We have some unfortunate issues with codecov and are evaluating how we will continue with the tool beyond the issues here. I can merge the PR with admin permissions once you think it is complete 👍

I have nothing to add, if you like, we can merge 👍
